permission_handler_android/lib/src/android.dart

```diff
@@ -0,0 +1,56 @@
+import 'package:flutter/foundation.dart';
+import 'package:permission_handler_android/src/permission_handler.pigeon.dart';
+
+import 'android_permission_handler_api_impls.dart';
+import 'android_object_mirrors/activity.dart';
+
+/// A static class that provided access to the attached Android Activity.
+///
+/// Usage:
+/// ```dart
+/// void main() {
+///   Android.onActivityChanged((Activity? activity) => this.activity = activity);
+/// }
+///
+/// void someMethod() {
+///   if (activity != null) {
+///     activity.getSystemService(Context.LOCATION_SERVICE);
+///
+///     ActivityCompat.shouldShowRequestPermissionRationale(
+///       activity,
+///       'permission_name',
+///     );
+///   }
+/// }
+/// ```
+class Android {
+  /// Hide constructor as this class should not be instantiated.
+  const Android._();
+
+  static ActivityFlutterApiImpl? _activityFlutterApi;
+
+  /// Reset this static class by clearing its reference to the
+  /// [ActivityFlutterApiImpl].
+  ///
+  /// For testing purposes only.
+  @visibleForTesting
+  static void reset() {
+    _activityFlutterApi = null;
+  }
+
+  /// Register the mirror APIs and obtain the attached Android Activity via the
+  /// [onActivityChanged] callback.
+  static void onActivityChanged(
+    void Function(Activity? activity) onActivityChanged, {
+    @visibleForTesting ActivityFlutterApiImpl? activityFlutterApi,
+  }) {
+    // Instantiate Activity Flutter API only once.
+    _activityFlutterApi ??= activityFlutterApi;
+    if (_activityFlutterApi == null) {
+      _activityFlutterApi = ActivityFlutterApiImpl();
+      ActivityFlutterApi.setup(_activityFlutterApi);
+    }
+
+    _activityFlutterApi!.registerOnActivityChanged(onActivityChanged);
+  }
```

Comment:
Wouldn't it be clearer to split this method up in a `onAttachedToActivity` and `onDetachedFromActivity`? 

Also I think we could add a `currentActivity` field or property which returns the current activity if one is attached. This would allow developers to get the current activity whenever they need it and don't have to rely on the `onAttachedToActivity` and `onDetachedFromActivity` methods.

This means we also don't have to immediately call the callbacks when they are registered.

Reply:
I am afraid that providing direct access to the `currentActivity` will be a source for bugs in consuming applications. Developers might not appreciate that the activity might become stale. With the current implementation, it is clear to the developers that the `Activity` is volatile, and they should program with this in mind.

For example: `Android.onActivityChanged` can be called by the developer directly, but is also called from `PermissionHandlerAndroid`. It can be unclear for developers whether the activity has attached already. For example, when depending solely on `permission_handler_android`, the developer needs to call `onAttachedToActivity`, as the suggested `currentActivity` will be `null`. If depending on `permission_handler` instead, `onAttachedToActivity` is already called by `PermissionHandlerAndroid`, and `currentActivity` **might** not be `null`, depending on when they fetch the activity. Concluding, it can be very opaque for developers how they should get ahold of an activity.

The way I see it, with the suggested approach, the developer would be tempted to write logic to tackle this uncertainty like this:

```dart
final Activity? activity = Android.currentActivity;
if (activity != null) {
  ...
} else {
  Android.onAttachedToActivity((Activity activity) => ...);
}
```